File: pdaily-server-postgresql/src/test/java/co/icesi/pdaily/business/structure/patient/PatientResourceIT.java
package co.icesi.pdaily.business.structure.patient;

import static io.restassured.RestAssured.given;
import static org.hamcrest.Matchers.*;

import java.util.List;

import javax.ws.rs.core.MediaType;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import co.haruk.core.testing.data.IDataSet;
import co.haruk.core.testing.data.IDataSetDependent;
import co.icesi.pdaily.business.structure.patient.app.PatientDTO;
import co.icesi.pdaily.testing.DataSets;
import co.icesi.pdaily.testing.PDailyTest;

@PDailyTest
@DisplayName("Patient tests")
public class PatientResourceIT implements IDataSetDependent {

	@Override
	public List<IDataSet> dataSets() {
		return List.of( DataSets.PATIENTS );
	}

	@Test
	@DisplayName("Finds all")
	void findByPatient() {
		given().contentType( MediaType.APPLICATION_JSON )
				.get( "/business/patient" )
				.then()
				.statusCode( 200 )
				.body(
						"size()", greaterThan( 0 )
				);
	}

	@Test
	@DisplayName("Finds by id correctly")
	void findById() {
		given().contentType( MediaType.APPLICATION_JSON )
				.get( "/business/patient/{0}", PatientTesting.PATIENT_ID )
				.then()
				.statusCode( 200 )
				.body(
						"id", equalToIgnoringCase( PatientTesting.PATIENT_ID ),
						"fullName", notNullValue()
				);
	}

	@Test
	@DisplayName("Saves a new patient")
	void savePatient() {
		final var dto = PatientDTO.of(
				null, "testing add"
		);
		given().contentType( MediaType.APPLICATION_JSON )
				.body( dto )
				.log().body()
				.post( "/business/patient" )
				.then()
				.log().body()
				.statusCode( 200 )
				.body(
						"id", notNullValue(),
						"fullName", equalToIgnoringCase( dto.fullName )
				);
	}

	@Test
	@DisplayName("Updates an patient")
	void updatesPatient() {
		final var dto = PatientDTO.of(
				PatientTesting.PATIENT_TO_UPDATE,
				"Testing update"
		);
		given().contentType( MediaType.APPLICATION_JSON )
				.body( dto )
				.post( "/business/patient" )
				.then()
				.statusCode( 200 )
				.body(
						"id", equalToIgnoringCase( dto.id ),
						"fullName", equalToIgnoringCase( dto.fullName )
				);
	}

	@Test
	@DisplayName("Deletes an patient")
	void deleteById() {
		given().contentType( MediaType.APPLICATION_JSON )
				.delete( "/business/patient/{0}", PatientTesting.PATIENT_TO_DELETE )
				.then()
				.statusCode( 204 );

		given().contentType( MediaType.APPLICATION_JSON )
				.get( "/business/patient/{0}", PatientTesting.PATIENT_TO_DELETE )
				.then()
				.statusCode( 404 );
	}
}

File: backend/backend/src/test/kotlin/co/pdaily/cloud/patients/boundary/PatientResourceIT.kt
package co.pdaily.cloud.patients.boundary

import co.pdaily.cloud.patients.PatientTesting
import co.pdaily.cloud.patients.entity.PatientDTO
import co.pdaily.cloud.patients.entity.PatientGender
import co.pdaily.cloud.testing.DataSets
import co.pdaily.cloud.testing.PdailyTest
import io.quarkus.test.junit.QuarkusTest
import org.junit.jupiter.api.DisplayName
import io.restassured.RestAssured.given
import io.restassured.http.ContentType
import org.hamcrest.Matchers
import org.junit.jupiter.api.Test

@QuarkusTest
@PdailyTest(sqlDataSets = [DataSets.PATIENTS])
@DisplayName("Patient tests")
internal class PatientResourceIT {

    @Test
    @DisplayName("Save a patient correctly")
    fun createDoctor(){
        val patient = PatientDTO(null,"112233", "Testing", 20,PatientGender.FEMENINO)
        given().body(patient)
                .contentType(ContentType.JSON)
                .`when`().post("/patients")
                .then()
                .assertThat()
                .statusCode(200)
                .and()
                .contentType(ContentType.JSON)
                .and()
                .body(
                        "id", Matchers.`is`(Matchers.notNullValue()),
                        "name", Matchers.equalToIgnoringCase(patient.name),
                        "document", Matchers.equalTo(patient.document),
                        "age", Matchers.equalTo(patient.age)
                )
    }

    @Test
    @DisplayName("Find all patients in DB")
    fun findAll(){
        given().`when`().get("/patients")
                .then()
                .log().all()
                .assertThat()
                .statusCode(200)
                .and()
                .contentType(ContentType.JSON)
                .and()
                .body(
                        "$.size", Matchers.greaterThan(0)
                )
    }

    @Test
    @DisplayName("Find patient by id")
    fun findById(){
        given().`when`()
                .get("/patients/{0}", PatientTesting.EXISTENT_ID)
                .then()
                .assertThat()
                .statusCode(200)
                .and().contentType(ContentType.JSON)
                .body("id", Matchers.equalToIgnoringCase(PatientTesting.EXISTENT_ID))
    }

    @Test
    @DisplayName("Delete patient by id")
    fun deleteById(){
        given().`when`().delete("/patients/{0}", PatientTesting.TO_DELETE_ID)
                .then()
                .log().all()
                .assertThat()
                .statusCode(204)

        given().`when`()
                .get("/patients/{0}", PatientTesting.TO_DELETE_ID)
                .then()
                .assertThat()
                .statusCode(500)
    }
}
